### PR TITLE
[MM-3996] Improved (add/join/leave/remove from channel/team) system messages 

### PR DESCRIPTION
**Pull request by saturninoabril**
#### Summary
- This PR combine all system messages of type "add/join/leave/remove from channel/team".
- These system messages will be marked as deleted at redux store with `"delete_at: 1, "state": "DELETE"`.
- Those consecutive system messages will combine into a single post with following new/modified fields: 
```
"message": "",
"type": "system_combined_user_activity",
"user_activity_posts": [post1, post2],
"system_post_ids": ["post_id_1", "post_id_2"],
```
- `system_post_ids` is used for deleting system post
- `user_activity_posts` is used for rendering on client side

Note: This will have breaking change on mobile RN since system messages will be deleted and combine post don'e have meaningful `message` value (it's empty). I'll submit separate PR for mobile RN.

#### Ticket Link
Jira ticket: [MM-3996](https://mattermost.atlassian.net/browse/MM-3996)

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] Added or updated unit tests (required for all new features)
- [x] All new/modified APIs include changes to the drivers

#### Test Information
This PR was tested on: [Chrome/FF, Mac OS] - to follow on mobile

### CODE REVIEW

File: src/reducers/entities/posts.js [hunk 1]
```diff
@@ -39,7 +39,10 @@ function handleReceivedPost(posts = {}, postsInChannel = {}, postsInThread = {},
         ];
     }
 
-    return {posts: nextPosts, postsInChannel: nextPostsInChannel, postsInThread: nextPostsInThread};
+    const combineSystemPosts = combineSystemPost(nextPostsInChannel[channelId], nextPosts);
```
Comment (hmhealey) [inline src/reducers/entities/posts.js]: When I looked at this yesterday, I wasn't really sure about combining system posts in the store. I was thinking it would be better to have the redux store match what is on the server, but then we'd have to remember to combine system posts and such all over the place before showing them in the UI, so I think this actually ends up being a really elegant solution to this

File: src/reducers/entities/posts.js [hunk 2]
```diff
@@ -145,13 +148,14 @@ function handleReceivedPosts(posts = {}, postsInChannel = {}, postsInThread = {}
 
     // Sort to ensure that the most recent posts are first, with pending
     // and failed posts first
-    postsForChannel.sort((a, b) => {
+    const filteredPostsForChannel = postsForChannel.filter((id) => nextPosts[id]).sort((a, b) => {
```
Comment (hmhealey) [inline src/reducers/entities/posts.js]: Why would we need to filter out post ids for posts that don't exist? Is there something that normally causes that to happen, or are you just being extra safe here?

File: src/reducers/entities/posts.js [hunk 3]
```diff
@@ -3,7 +3,7 @@
 
 import {PostTypes, SearchTypes, UserTypes} from 'action_types';
 import {Posts} from 'constants';
-import {comparePosts} from 'utils/post_utils';
+import {comparePosts, combineSystemPost} from 'utils/post_utils';
```
Comment (hmhealey) [inline src/reducers/entities/posts.js]: This method should be `combineSystemPosts` so that it's clearer that it does all posts at once

File: src/utils/helpers.js [hunk 4]
```diff
@@ -79,3 +79,24 @@ export const sendMessageToRemoveBackwardsCompatibility = (currentVersion, minVer
         console.warn('You can now remove backwards-compatibility code from the project, as referenced in the stacktrace below.'); //eslint-disable-line no-console
     }
 };
+
+// Generates a RFC-4122 version 4 compliant globally unique identifier.
+export function generateId() {
```
Comment (hmhealey) [inline src/utils/helpers.js]: Can you make the `TestHelper.generateId` method in `test/test_helper.js` call this so that we're not duplicating the logic?

File: src/utils/post_utils.js [hunk 5]
```diff
@@ -175,3 +183,98 @@ export function comparePosts(a, b) {
...
+            }
+        }
+
+        if (channelPost.type === '') {
+            if (userActivitySystemPosts.length > 0) {
+                const combinedPost = {
+                    id: combinedPostId || generateId(),
+                    root_id: '',
+                    type: Posts.POST_TYPES.COMBINED_USER_ACTIVITY,
+                    message: '',
```
Comment (hmhealey) [inline src/utils/post_utils.js]: You mentioned that having no message will make this not work with the mobile app temporarily, but what if you make the combined post's message just the messages of each system posts joined together? That would help solve that case and also make it easier if anyone else is building a client with this

File: src/utils/post_utils.js [hunk 6]
```diff
@@ -175,3 +183,98 @@ export function comparePosts(a, b) {
...
+            postsForChannel.push(channelPost.id);
+        } else if (
+                userActivitySystemPosts.length === MAX_COMBINED_SYSTEM_POSTS ||
+                (userActivitySystemPosts.length > 0 && i === postsIds.length - 1)
+            ) {
+            const combinedPost = {
+                id: combinedPostId || generateId(),
+                root_id: '',
+                type: Posts.POST_TYPES.COMBINED_USER_ACTIVITY,
+                message: '',
```
Comment (hmhealey) [inline src/utils/post_utils.js]: Same here

File: src/utils/post_utils.js [hunk 7]
```diff
@@ -175,3 +183,98 @@ export function comparePosts(a, b) {
...
+            } else if (channelPost.type === Posts.POST_TYPES.COMBINED_USER_ACTIVITY) {
+                userActivitySystemPosts = [...userActivitySystemPosts, ...channelPost.user_activity_posts];
+                systemPostIds = [...systemPostIds, ...channelPost.system_post_ids];
+
+                combinedPostId = channelPost.id;
+            }
+        }
+
+        if (channelPost.type === '') {
+            if (userActivitySystemPosts.length > 0) {
```
Comment (hmhealey) [inline src/utils/post_utils.js]: It would be nice if we could combine the two branches of this if statement since their logic is almost identical. Would this work if it was something like
```javascript
if (
    (channelPost.type === '' && userActivitySystemPosts.length > 0) ||
    userActivitySystemPosts.length === MAX_COMBINED_SYSTEM_POSTS ||
    (userActivitySystemPosts.length > 0 && i === postsIds.length - 1)
) {
    ...
}
```

File: src/reducers/entities/posts.js (hunk 2 above)
Comment (saturninoabril) [reply src/reducers/entities/posts.js]: I'm seeing `undefined` after deleting combined post; somehow the ID of combined post remained in the postsForChannel.  Not sure what I'm missing there but that's something I need to investigate further.

File: src/utils/post_utils.js (hunk 5 above)
Comment (saturninoabril) [reply src/utils/post_utils.js]: I think one of my worry is the localization but I'll see what I can do here.

Comment (hmhealey) [reply src/utils/post_utils.js]: I think it's fine if it's a simpler message like "Harrison joined the channel. Harrison left the channel." here, but it'll make sure that we have something to fall back to

File: src/reducers/entities/posts.js (hunk 2 above)
Comment (saturninoabril) [reply src/reducers/entities/posts.js]: Fixed this by adding `channel_id` to combined post.

File: src/utils/helpers.js (hunk 4 above)
Comment (saturninoabril) [reply src/utils/helpers.js]: Done

File: src/utils/post_utils.js (hunk 5 above)
Comment (saturninoabril) [reply src/utils/post_utils.js]: hmmm, forgot about this... will update after PM review.

File: src/utils/post_utils.js (hunk 7 above)
Comment (saturninoabril) [reply src/utils/post_utils.js]: Done

File: src/reducers/entities/posts.js [hunk 8]
```diff
@@ -39,7 +39,10 @@ function handleReceivedPost(posts = {}, postsInChannel = {}, postsInThread = {},
         ];
     }
 
-    return {posts: nextPosts, postsInChannel: nextPostsInChannel, postsInThread: nextPostsInThread};
+    const withCombineSystemPosts = combineSystemPosts(nextPostsInChannel[channelId], nextPosts);
+    nextPostsInChannel[channelId] = withCombineSystemPosts.postsForChannel;
+
+    return {posts: withCombineSystemPosts.nextPosts, postsInChannel: nextPostsInChannel, postsInThread: nextPostsInThread};
```
Comment (enahum) [inline src/reducers/entities/posts.js]: `nextPostsInChannel` and `nextPostsInThread` here should have the ids for those posts that were removed from the combine stuff as you will end up having something like this for example: for 3 system messages in a channel you'll have 1 post in the posts entity and 3 posts in the postsInChannel entity.

sounds about right? probably the cause that you needed to filter later on

File: src/reducers/entities/posts.js [hunk 9]
```diff
@@ -149,9 +152,10 @@ function handleReceivedPosts(posts = {}, postsInChannel = {}, postsInThread = {}
         return comparePosts(nextPosts[a], nextPosts[b]);
     });
 
-    nextPostsInChannel[channelId] = postsForChannel;
+    const withCombineSystemPosts = combineSystemPosts(postsForChannel, nextPosts, channelId);
+    nextPostsInChannel[channelId] = withCombineSystemPosts.postsForChannel;
 
-    return {posts: nextPosts, postsInChannel: nextPostsInChannel, postsInThread: nextPostsInThread};
+    return {posts: withCombineSystemPosts.nextPosts, postsInChannel: nextPostsInChannel, postsInThread: nextPostsInThread};
```
Comment (enahum) [inline src/reducers/entities/posts.js]: same as before

File: src/selectors/entities/posts.js [hunk 10]
```diff
@@ -230,7 +230,9 @@ export function makeGetPostsInChannel() {
                 }
 
                 const previousPost = allPosts[postIds[i + 1]] || {create_at: 0};
-                posts.push(formatPostInChannel(post, previousPost, i, allPosts, postsInThread, postIds, currentUser));
+                if (post) {
```
Comment (enahum) [inline src/selectors/entities/posts.js]: I don't mind being defensive here at all, but out of curiosity why did you need it?

File: src/utils/helpers.js (hunk 4 above)
Comment (enahum) [reply src/utils/helpers.js]: are you going to be removing this helper?

File: src/utils/post_utils.js [hunk 11]
```diff
@@ -3,10 +3,14 @@
...
 import {General, Posts, Preferences, Permissions} from 'constants';
 
-import {getPreferenceKey} from './preference_utils';
 import {hasNewPermissions} from 'selectors/entities/general';
 import {haveIChannelPermission} from 'selectors/entities/roles';
 
+import {generateId} from './helpers';
+import {getPreferenceKey} from './preference_utils';
+
+const MAX_COMBINED_SYSTEM_POSTS = 100;
```
Comment (enahum) [inline src/utils/post_utils.js]: curiosity: why 100? who came up with this number and why?

File: src/utils/post_utils.js [hunk 12]
```diff
@@ -175,3 +183,82 @@ export function comparePosts(a, b) {
...
+        if (
+            channelPost.delete_at === 0 &&
+            (isUserActivityPost(channelPost.type) || channelPost.type === Posts.POST_TYPES.COMBINED_USER_ACTIVITY)
+        ) {
+            if (!createAt || createAt > channelPost.create_at) {
+                createAt = channelPost.create_at;
+            }
+
+            if (isUserActivityPost(channelPost.type)) {
+                userActivitySystemPosts = [...userActivitySystemPosts, channelPost];
```
Comment (enahum) [inline src/utils/post_utils.js]: why are you creating a new array every time instead of just adding the post to the existing array?

File: src/utils/post_utils.js [hunk 13]
```diff
@@ -175,3 +183,82 @@ export function comparePosts(a, b) {
...
+                systemPostIds.push(channelPost.id);
+
+                if (nextPosts[channelPost.id]) {
+                    nextPosts[channelPost.id] = {...channelPost, state: Posts.POST_DELETED, delete_at: 1};
+                }
+            } else if (channelPost.type === Posts.POST_TYPES.COMBINED_USER_ACTIVITY) {
+                userActivitySystemPosts = [...userActivitySystemPosts, ...channelPost.user_activity_posts];
+                systemPostIds = [...systemPostIds, ...channelPost.system_post_ids];
```
Comment (enahum) [inline src/utils/post_utils.js]: same here and the line above

File: src/utils/post_utils.js [hunk 14]
```diff
@@ -175,3 +183,82 @@ export function comparePosts(a, b) {
...
+            createAt = null;
+            combinedPostId = null;
+
+            if (channelPost.type === '') {
+                postsForChannel.push(channelPost.id);
+            }
+        } else if (channelPost.type === '') {
+            postsForChannel.push(channelPost.id);
+        }
+    });
```
Comment (enahum) [inline src/utils/post_utils.js]: after adding the new combined post to the `postsForChannel` you should remove those that were marked as `DELETED`? or do you see a better way to accomplish that?

File: src/reducers/entities/posts.js (hunk 8 above)
Comment (saturninoabril) [reply src/reducers/entities/posts.js]: Got it, I'll update this.

File: src/selectors/entities/posts.js (hunk 10 above)
Comment (saturninoabril) [reply src/selectors/entities/posts.js]: Ah, i missed to revert this. This change is not needed anymore. (The problem I'm seeing initially is already fixed. It was fixed by adding `channel_id` in the combinedPost)

File: src/utils/post_utils.js (hunk 12 above)
Comment (saturninoabril) [reply src/utils/post_utils.js]: It's only out of convenience during my first iteration :) and just missed to clean up. Will update using push instead.

File: src/utils/post_utils.js (hunk 14 above)
Comment (saturninoabril) [reply src/utils/post_utils.js]: I'm just marking those system posts as deleted so that for the next round of received post/s those are known as received already and no need to merge with existing combinedPost. I think it's straightforward rather than checking each time whether the combinedPost has already the system post or not.

File: src/utils/helpers.js (hunk 4 above)
Comment (saturninoabril) [reply src/utils/helpers.js]: Not on this PR, but after this has been merged. I'll clean up that code - will delete `TestHelper.generateId()` and will just directly call `UtilsHelper.generateId()` for our test files.

(I initially thought to consider that in this PR but there are too many `TestHelper.generateId()` everywhere and I find it unnecessary for this PR.)

File: src/reducers/entities/posts.js (hunk 8 above)
Comment (saturninoabril) [reply src/reducers/entities/posts.js]: Can confirm that those system posts are included in `nextPostsInChannel` but not necessary in `nextPostsInThread`. Those system posts on `nextPostsInChannel` were marked as deleted.

**Comment by saturninoabril** [comment]
@hmhealey @enahum I've updated the PR so that this will not break the current Mobile RN (like the screenshots below just in case this PR gets merged without change on it yet). I'm currently working on Mobile RN and will submit later.

screen shot 2018-05-14 at 9 45 54 pm
screen shot 2018-05-14 at 9 44 58 pm
screen shot 2018-05-14 at 9 45 12 pm

File: src/actions/posts.js [hunk 15]
```diff
@@ -247,23 +247,36 @@ export function resetCreatePostRequest() {
 }
 
 export function deletePost(post) {
-    return async (dispatch) => {
+    return async (dispatch, getState) => {
         const delPost = {...post};
+        if (delPost.type === Posts.POST_TYPES.COMBINED_USER_ACTIVITY) {
+            delPost.system_post_ids.forEach((systemPostId) => {
+                const systemPost = Selectors.getPost(getState(), systemPostId);
```
[the remark this answers is not in the document]
Comment (saturninoabril) [reply src/actions/posts.js]: Thanks @enahum! You're right. I've made an update into it.

[the remark this answers is not in the document]
Comment (saturninoabril) [reply src/actions/posts.js]: Thanks @enahum! You're right. I've made an update into it.
